throughput/spam.go

```diff
@@ -0,0 +1,427 @@
+		select {
+		case <-it.C:
+			start := time.Now()
+
+			// Check to see if we should wait for pending txs
+			if int64(currentTarget)+inflight.Load() > int64(currentTarget*pendingTargetMultiplier) {
+				consecutiveUnderBacklog = 0
+				consecutiveAboveBacklog++
+				if consecutiveAboveBacklog >= failedRunsToDecreaseTarget {
+					if currentTarget > s.txsPerSecondStep {
+						currentTarget -= s.txsPerSecondStep
+						utils.Outf("{{cyan}}skipping issuance because large backlog detected, decreasing target tps:{{/}} %d\n", currentTarget)
+					} else {
+						utils.Outf("{{cyan}}skipping issuance because large backlog detected, cannot decrease target{{/}}\n")
+					}
+					consecutiveAboveBacklog = 0
+				}
+				it.Reset(1 * time.Second)
+				break
+			}
+
+			// Issue txs
+			g := &errgroup.Group{}
+			g.SetLimit(maxConcurrency)
+			for i := 0; i < currentTarget; i++ {
+				senderIndex, recipientIndex := z.Uint64(), z.Uint64()
+				sender := accounts[senderIndex]
+				if recipientIndex == senderIndex {
+					if recipientIndex == uint64(s.numAccounts-1) {
+						recipientIndex--
+					} else {
+						recipientIndex++
+					}
+				}
+				recipient := accounts[recipientIndex].Address
+				issuer := getRandomIssuer(issuers)
+				g.Go(func() error {
+					factory := factories[senderIndex]
+					fundsL.Lock()
+					balance := funds[sender.Address]
+					if feePerTx > balance {
+						fundsL.Unlock()
+						utils.Outf("{{orange}}tx has insufficient funds:{{/}} %s\n", sender.Address)
+						return fmt.Errorf("%s has insufficient funds", sender.Address)
+					}
+					funds[sender.Address] = balance - feePerTx - amountToTransfer
+					funds[recipient] += amountToTransfer
+					fundsL.Unlock()
+
+					// Send transaction
+					actions := sh.GetTransfer(recipient, amountToTransfer, uniqueBytes())
+					return issuer.Send(cctx, actions, factory, feePerTx)
+				})
+			}
+
+			// Wait for txs to finish
+			if err := g.Wait(); err != nil {
+				// We don't return here because we want to return funds
+				utils.Outf("{{orange}}broadcast loop error:{{/}} %v\n", err)
+				stop = true
+				break
+			}
+
+			// Determine how long to sleep
+			dur := time.Since(start)
+			sleep := max(float64(consts.MillisecondsPerSecond-dur.Milliseconds()), 0)
+			it.Reset(time.Duration(sleep) * time.Millisecond)
+
+			// Check to see if we should increase target
+			consecutiveAboveBacklog = 0
+			consecutiveUnderBacklog++
+			if consecutiveUnderBacklog >= successfulRunsToIncreaseTarget && currentTarget < s.txsPerSecond {
+				currentTarget = min(currentTarget+s.txsPerSecondStep, s.txsPerSecond)
+				utils.Outf("{{cyan}}increasing target tps:{{/}} %d\n", currentTarget)
+				consecutiveUnderBacklog = 0
+			} else if currentTarget == s.txsPerSecond && consecutiveUnderBacklog >= successfulRunsToIncreaseTarget {
+				if terminate {
+					utils.Outf("{{green}}reached target tps:{{/}} %d\n", currentTarget)
+					// Cancel the context to stop the issuers
+					cancel()
+				}
+			}
```

Comment:
Could we add `terminate` as the first condition in the `if else` branch and add a comment here explaining why this was added (I assume to terminate once hitting desired TPS for the e2e tests) ?